### PR TITLE
[Perf] Reference audio cache for MOSS-TTS-Local-v1.5 (LRU + single-flight dedup)

### PR DESCRIPTION
**Pull request by JiaxinD**
# [Perf] Reference audio cache for MOSS-TTS-Local-v1.5 (LRU + single-flight dedup)

Addresses #731 (roadmap: #637) — the MOSS-TTS reference-audio encoder LRU cache.  
Builds on AkazaAkane's prior design in #648 and the Higgs reference cache in #605.

---

## Motivation

MOSS-TTS-Local-v1.5's `_BatchedReferenceEncoder` runs the ~1B-param causal codec encoder
for every reference audio, costing roughly **0.25 GPU-seconds per encode** on the codec
device. #728 already coalesces concurrent requests into batches (up to 8 per 4 ms window,
`stages.py:202`), but the same reference arriving at different times still re-encodes
independently. For the common "fixed speaker" workload this is redundant work on a GPU
already shared with AR prefill.

Radix-cache in the SGLang backend covers AR-side prefix reuse (#728 body); this PR targets
only the **encode side** (the incremental cost above and beyond what #728 provides).

---

## What this PR does (5 commits)

| Commit | Title | Key change |
|---|---|---|
| a5fa07c | `[Refactor]` Extract shared reference-audio cache key helpers | Higgs memo/sentinel/hash functions → `preprocessing/cache_key.py`; Higgs re-exports aliases (zero behaviour change) |
| 400e189 | `[Perf]` Add cached reference encoder for MOSS-TTS Local | `CachedReferenceEncoder` wrapping `_BatchedReferenceEncoder` |
| 6a7b7b1 | `[Config]` Wire ref-audio cache flags into preprocessing executor | Three kwargs + `config.py` factory_args; 60 s stats logging |
| b19baa7 | `[Perf]` Cache data-URI references (bytes: keyspace) + add duration gate | `encode_data_uri`; `bytes:` keyspace; data-URI duration check |
| fa4c22a | `[Perf]` Add opt-in stat-trusting fast path for reference cache keys | `trust_stat` flag (zero-I/O memo hit); co-authored with GaokaiZhang (#740) |
| 2855ac1 | `[Review]` Address review | validate capacity; DRY `_cached_encode`; **MOSS defaults `trust_stat=False`**; log cadence constant |

---

## Cross-model note (commit a5fa07c touches `higgs_tts/stages.py`)

Commit 1 is a pure **extraction refactor** of `higgs_tts/stages.py`: the memo dict,
sentinel reader, and `reference_path_cache_key` function were moved verbatim into
`sglang_omni/preprocessing/cache_key.py`, and `higgs_tts/stages.py` now re-imports them
at the same names (`stages.py:53–60`). No Higgs logic was changed; all existing Higgs
tests pass without modification (`tests/unit_test/higgs_tts/test_pipeline.py:117–194`).
The refactor is a prerequisite so MOSS-TTS-Local can share the same key scheme without
duplicating 80 lines.

---

## Architecture

```
request
  └─ CachedReferenceEncoder.encode(path)
        │
        ├─ 1. duration gate (>100 s → reject)           stages.py:266
        │     (file path: gated first via cheap torchaudio.info;
        │      data-URI: checked inside the leader so concurrent
        │      same-payload requests share one sf.read)
        ├─ 2. key = reference_path_cache_key(path)       stages.py:267
        │       └─ stat-memo (size+mtime+ctime) → sentinel (8 KB×3) → xxh3_64
        │          → "file:{hex}" or None (missing/URL)  cache_key.py:115–135
        │
        ├─ 3. cache hit → return stored.clone().to(long) stages.py:276–280
        ├─ 4. in-flight (single-flight) → follower waits stages.py:280–286
        └─ 5. miss → register leader Future
                └─ _BatchedReferenceEncoder.encode(path) (4 ms coalescer)
                        └─ processor.encode_audios_from_path(unique_paths)
                              ↓
               TOCTOU re-stat before put               stages.py:313–319
               cache.put(key, stored_int32_cpu)
               future.set_result(stored)
```

**Keyspaces** (two distinct prefixes, never cross-contaminate):

| Input | Prefix | Key content |
|---|---|---|
| File path | `file:` | xxh3\_64 of full file bytes, gated by stat-memo + sentinel |
| data-URI | `bytes:` | xxh3\_64 of raw decoded bytes |
| Missing / URL / tensor | `None` | cache and single-flight bypassed entirely |

**Storage**: `detach().to("cpu", torch.int32)` — matches Higgs code-cache discipline
(`higgs_tts/stages.py:365`). On hit: `.clone().to(torch.long)`. On miss: original codec
tensor returned unchanged (bit-identical to OFF=True path).

**Capacity defaults** (`config.py:23–25`):

| Flag | Default | Worst-case RAM |
|---|---|---|
| `ref_audio_cache_max_items` | 256 | ~15 MB (100 s clips × 12 VQ × 4 B × 256) |
| `ref_audio_cache_max_bytes` | 64 MiB | hard backstop |

**Failure semantics**: leader failure clears `_inflight` without writing to cache;
each follower receives an independent `RuntimeError(...) from cause` (not the shared
exception instance — mirrors the existing `stages.py:216–221` note in
`_BatchedReferenceEncoder`). Third request becomes a new leader and may succeed.

---

## Correctness gate

Unit test `test_cached_reference_encoder_on_off_hit_bit_identical` (`test_pipeline.py:~490`):
- Runs the full `_prepare_moss_tts_local_request` chain with a fake-but-correct processor
- Asserts OFF path == ON-miss path == ON-hit path (identical `prompt_rows` and
  `input_ids_list`)
- Verifies int32 round-trip is lossless at the boundary value 1023
- Confirms `encode_count == 1` on the second (hit) call

Smoke test on novita-h100 with real MOSS-TTS-Local-Transformer-v1.5 weights (GPU 3):
- hit-path CPU tensor accepted by `processor.build_user_message` ✓
- T=62 frames / 5.0 s = 12.40 Hz ≈ 12.5 Hz empirically confirmed ✓

---

### Corroborating numbers from #740 (same approach, GaokaiZhang's implementation)

> These measure #740's implementation of the same content-keyed LRU, not this
> branch's code. They corroborate the mechanism; our own Phase 3 numbers on this
> branch (table below) will be the primary evidence. Source: #740 PR body, 2×H100.

- Hit-path effect (same reference, sequential requests, live server): first request
  (miss) **2.44 s** → subsequent requests (hits) **0.81 s** mean.
- SeedTTS EN full (pure miss path, every reference unique), c=16, 2×H100:
  rtf_mean **0.4811**, WER 2.27% raw / 1.71% excl., 1088/1088, 0 failed, 7.91 req/s.
  Same-day baseline on merged main: rtf_mean 0.5053, WER 2.21% raw / 1.79% excl.
  (deltas within the run-to-run band documented in #728).

### Our Phase 3 (this branch) — measured

> **Environment (full disclosure):** novita-h100. A sole-occupant run was planned on NUMA1
> (GPU5+6); mid-run an unrelated co-tenant took GPU6+7 and saturated NUMA1 CPU, so the
> benchmark **migrated to GPU4+5 with CPU pinned to NUMA0** (`numactl --cpunodebind=0
> --membind=0`). Cross-NUMA GPU access is a **constant** overhead identical for the ON and
> OFF arms and every round (ABAB interleaved), so it does **not bias the A/B** — it only
> shifts the absolute rtf baseline. Each round ran a co-tenant-contamination gate; all 6
> Load-A rounds passed clean. Full methodology + per-round data in the run report.

**Load A — SeedTTS EN full set (1088), ABAB ON/OFF ×3 (1088/1088 each, 0 failed)**

| metric | Cache OFF (mean ± 95% CI) | Cache ON (mean ± 95% CI) |
|---|---|---|
| RTF mean | 0.5096 ± 0.0387 | 0.4973 ± 0.0257 |
| throughput req/s | 7.443 ± 0.633 | 7.637 ± 0.414 |
| RTF p99 | 1.0547 ± 0.1942 | 1.0029 ± 0.0450 |

The ON/OFF 95% CIs **overlap** → the cache is statistically indistinguishable from OFF =
**no regression** (zero measurable overhead). The nominal ~2.5% ON edge is within noise at 3
rounds; the saved encode work surfaces as returned GPU headroom (below), not an rtf gain at
c=16 where AR generation dominates.

**Native reuse + savings (hard counts).** SeedTTS EN recurs speakers across target texts, so
the standard benchmark itself exercises the cache. One full-set ON snapshot:

| lookups | encodes (misses) | hits | merged (single-flight) | reuse | encodes avoided | GPU-s saved |
|---|---|---|---|---|---|---|
| 915 | 566 | 71 | 278 | 38.1% | 349 | ~87 |

`merged` (single-flight dedup) is a capability **#740 lacks** — hundreds of concurrent
same-reference requests folded onto one in-flight encode instead of re-encoding.

**WER sanity (quality non-regression):** ON **2.39% raw / 1.82% excl-outliers** vs OFF
**2.22% raw / 1.68% excl** (Qwen3-ASR, 1088/1088 each, 6 outliers each). The ~0.17%
difference is AR sampling noise (temp=0.8 → non-identical audio run-to-run) — same band as
#740's 2.27%/2.21%. Codes-level ON==re-encode identity is separately proven by unit test
T5; WER_ON ≈ WER_OFF confirms no audible regression.

**Load B — fixed-speaker reuse sweep (N=512, cache ON), mechanism curve (hard counts):**

| K speakers | encodes performed (=K) | encodes saved (512−K) | GPU-s saved (×0.25) | hit-rate |
|---|---|---|---|---|
| 1 | 1 | 511 | ~128 | 0.998 |
| 4 | 4 | 508 | ~127 | 0.992 |
| 16 | 16 | 496 | ~124 | 0.969 |
| 64 | 64 | 448 | ~112 | 0.875 |
| 256 | 256 | 256 | ~64 | 0.500 |

Lower K → more reuse → more encodes skipped → more GPU-seconds returned to the AR engine.
(Load-B end-to-end throughput was AR-generation-bound with token-count uncontrolled in the
client, so it is **not** a cache comparison; the value is the exact encode/GPU-second
savings above, which are hard counts with zero statistical noise.)

---

## Relation to #740 and #730

**#740** ("[Perf] MOSS-TTS Local: content-keyed LRU for reference codes", GaokaiZhang)
independently implemented the same #637 roadmap item. We have **converged on this PR
as the base** (coordinated with GaokaiZhang). The stat-only memo fast path from #740 is
incorporated here as the opt-in `trust_stat` flag (**MOSS defaults it `False` per review;
opt-in for immutable-reference deployments**) (commit "Add opt-in stat-trusting fast
path", co-authored with GaokaiZhang), and #740's benchmark numbers are cited below with
attribution. #740 will close in favor of this PR.

**#730 / #744** (decouple): #744 ("Split preprocessing; add audio_encoder stage") is the
MOSS-TTS-Local half of the #730 decouple — it moves the heavy reference-encode work out of
`preprocessing` into a dedicated `audio_encoder` stage. This PR is **complementary**: the
cache is an outer wrapper around that encoder and migrates with it. When #744 merges after
#743, the `CachedReferenceEncoder` wrapping and its flags move from
`create_preprocessing_executor` to `create_audio_encoder_executor` (~13 lines, no logic
change); reference-encode semantics are byte-identical across the split. Porting the same
cache to the older `moss_tts` (Delay) path is a separate follow-up (cf. AkazaAkane's
#648/#699).

---

## Cap discussion / Future work

The reference encode cache and the AR-side prefix cache are **orthogonal**: a cache hit
eliminates the codec GPU-second but still consumes a preprocessing-worker slot from
`max_concurrency=16`. If the AR concurrency cap is raised to 32 in a future PR,
`preprocessing max_concurrency=16` may become the new bottleneck under high load —
the cache shifts where time is spent (codec → Python dispatch) but does not by itself
increase the worker pool size. Raising `max_concurrency` in tandem, or combining with
the async-decode lookahead work in the pipeline, would complete the optimisation.

---

## Testing

```
pytest tests/unit_test/moss_tts_local/test_pipeline.py  -x   # T1–T9 + existing tests
pytest tests/unit_test/higgs_tts/test_pipeline.py       -x   # Higgs regression (T1/T2)
pytest tests/unit_test/preprocessing/                   -x   # key helper tests
```

94 collected, 94 passed on novita-h100 (sglang-omni docker, full deps).

New tests added:

| Test | What it covers |
|---|---|
| `test_cached_reference_encoder_on_off_hit_bit_identical` | Bit-identical gate (T5) |
| `test_cached_reference_encoder_lru_eviction` | Item-count + byte-budget eviction (T3) |
| `test_cached_reference_encoder_true_concurrency_dedup` | Single-flight, 8 threads (T4) |
| `test_cached_reference_encoder_failure_does_not_poison` | Independent exceptions, no cache poison (T6) |
| `test_cached_reference_encoder_return_value_isolation` | Clone isolation (T7) |
| `test_cached_reference_encoder_duration_gate` | >100 s rejected pre-inflight (T8) |
| `test_cached_reference_encoder_data_uri_hit_miss` | bytes: keyspace hit/miss |
| `test_cached_reference_encoder_file_bytes_keyspaces_do_not_collide` | Keyspace isolation |
| `test_cached_reference_encoder_data_uri_duration_gate` | data-URI duration check |

### CODE REVIEW

File: sglang_omni/models/moss_tts_local/stages.py [hunk 1]
```diff
@@ -227,6 +233,206 @@ def _worker(self) -> None:
...
+                    f"reference encode failed (data-URI, merged): {cause}"
+                ) from cause
+            self._maybe_log()
+            return stored.clone().to(torch.long)
+
+        assert leader_fut is not None
+        try:
+            import soundfile as sf
+
+            audio, sample_rate = sf.read(
```
Comment (gaoyang07) [inline sglang_omni/models/moss_tts_local/stages.py:396]: We need to read raw first, and then decode the raw to avoid invalid _DATA_URI_RE refs?

File: sglang_omni/models/moss_tts_local/stages.py [hunk 2]
```diff
@@ -227,6 +233,206 @@ def _worker(self) -> None:
...
+            *snapshot,
+        )
+
+    def encode_data_uri(self, ref_audio: str, *, processor: Any) -> torch.Tensor:
+        """Cache-aware encode for data-URI refs through the same LRU + single-flight
+        as file paths (adds the duration check _reference_for_processor lacks).
+
+        Note(Jiaxin): file: and bytes: keyspaces never collide — the two decode
+        chains differ, so codes aren't guaranteed identical for the "same" audio.
+        """
```
Comment (Ccyest) [inline sglang_omni/models/moss_tts_local/stages.py:379]: encode() and encode_data_uri() each implement a nearly identical single-flight flow.

Suggestion: Extract a private method _cached_encode(key, encode_fn) that encapsulates the single-flight skeleton. encode and encode_data_uri would only be responsible for key computation + constructing the encode_fn callable.

File: sglang_omni/models/moss_tts_local/stages.py [hunk 3]
```diff
@@ -227,6 +233,206 @@ def _worker(self) -> None:
...
+            if rekey == key:
+                self._cache.put(key, stored)
+            self._inflight.pop(key, None)
+        leader_fut.set_result(stored)
+        self._maybe_log()
+        return result  # original tensor: miss path stays bit-identical to cache-off
+
+    def _maybe_log(self) -> None:
+        now = time.monotonic()
+        if now - self._last_log_time < 60.0:
```
Comment (gaoyang07) [inline sglang_omni/models/moss_tts_local/stages.py:355]: Make 60.0 configurable maybe?

File: sglang_omni/models/moss_tts_local/stages.py [hunk 4]
```diff
@@ -227,6 +233,206 @@ def _worker(self) -> None:
...
+        self._last_log_time: float = 0.0
+
+    def encode(self, path: str) -> torch.Tensor:
+        path = str(path)
+        # Note(Jiaxin): duration gate runs first — a >100 s ref must never reach
+        # the cache or the inflight dict.
+        _BatchedReferenceEncoder._check_reference_duration(path)
+        # trust_stat: stat-only fast path; the codes lookup needs the key on every
+        # request (see reference_path_cache_key).
+        key = _reference_path_cache_key(path, trust_stat=True)
```
Comment (gaoyang07) [inline sglang_omni/models/moss_tts_local/stages.py]: Is it ok to set `trust_stat=True` as default?

File: sglang_omni/models/moss_tts_local/stages.py [hunk 5]
```diff
@@ -227,6 +233,206 @@ def _worker(self) -> None:
...
+    Hit path returns a fresh .clone().to(long) so callers cannot mutate cached state.
+    Stores codes as int32 on CPU (lossless for codebook values in [0, 1023]).
+    """
+
+    def __init__(
+        self,
+        encoder: _BatchedReferenceEncoder,
+        *,
+        max_items: int = 256,
+        max_bytes: int = 64 * 1024 * 1024,
```
Comment (Hayden727) [inline sglang_omni/models/moss_tts_local/stages.py:252]: [P3] These are new public `factory_args`, so it would be safer to validate them before building the `StageOutputCache`. Today a negative `ref_audio_cache_max_items` starts successfully, but the first cacheable `put()` can pop the only entry and then keep evicting from an empty `OrderedDict`, raising `KeyError` at request time. Can we fail fast here (or in `create_preprocessing_executor`) for negative capacities and add a small negative-config regression test?

File: sglang_omni/models/moss_tts_local/stages.py (hunk 1 above)
Comment (AkazaAkane) [inline sglang_omni/models/moss_tts_local/stages.py:396]: One minor note: the PR mentions that the duration gate happens before in-flight registration, but the code actually registers _inflight[key] before running sf.read() and the duration check. I actually think the code's way is better, as it prevents concurrent sf.read() calls on the same massive payload (saving CPU/RAM)

File: sglang_omni/models/moss_tts_local/stages.py (hunk 4 above)
Comment (AkazaAkane) [reply sglang_omni/models/moss_tts_local/stages.py]: trust_stat=True -> Once a file has been hashed once, future requests avoid even the small sentinel read. That matters because the cache key is computed on every request.

But if the path is reused as a mutable scratch file, （for example /tmp/ref.wav）and a caller overwrites it between requests, the cache may return stale codes if the stat tuple does not change enough. The worst outcome is not a crash; it is incorrect speaker/reference conditioning.

So probably need 1. a comment/config option saying MOSS assumes path references are immutable while serving; or 2. A safe-mode option using trust_stat=False for deployments with mutable reference paths.

File: sglang_omni/models/moss_tts_local/stages.py [hunk 6]
```diff
@@ -227,6 +233,206 @@ def _worker(self) -> None:
...
+
+        from sglang_omni.models.moss_tts.request_builders import _DATA_URI_RE
+        from sglang_omni.preprocessing.cache_key import hash_bytes as _hash_bytes
+
+        match = _DATA_URI_RE.match(ref_audio)
+        if match is None:
+            raise ValueError(f"encode_data_uri: not a data URI ({ref_audio[:40]!r}...)")
+
+        raw = base64.b64decode(match.group("data"))
+        key = f"bytes:{_hash_bytes(raw)}"
```
Comment (Ccyest) [inline sglang_omni/models/moss_tts_local/stages.py:391]: encode_data_uri decodes the entire data-URI payload (potentially hundreds of KB) before computing hash_bytes(raw) for the cache key. On a cache hit, these decoded raw bytes are immediately discarded, each time can cause CPU waste.

I feel like hashing the base64-encoded string directly as the cache key can be a solution, but I checked higgs we also did the entire data hashing. I remember we might discussed it previously but forgot why we made that decision. 

Theoretically it should be correct tho, any thoughts?

Comment (JiaxinD) [reply sglang_omni/models/moss_tts_local/stages.py:391]: Dug up the original decision, it's from #605 (commit 1bbac5a, by BBuf), recorded in the tests rather than a discussion thread: test_higgs_reference_cache_key_ignores_media_type pins down two invariants: (1) media_type is a decode hint the codec ignores, so it must not split keys; (2) raw bytes and their base64 encoding must resolve to the same content key. Hashing the base64 string would violate both: the same audio arriving as bytes vs base64, or wrapped with a different media_type, would miss the cache and re-encode. That's why the key is computed on decoded raw bytes (content-addressed): _reference_audio_cache_key, then b64decode, then hash_media_item(raw), giving bytes:{hash}.

On skipping the decode: I looked at the numbers and it's thinner than it sounds. The expensive part (0.25s codec encode) is already absorbed by the codes cache on a hit, so what's left on the hit path is just the b64decode + xxh3, both µs-level C code. A string-level memo in front (same structure as our _REF_PATH_HASH_MEMO on the file path) is ~1-2h to add and low-risk, but the memo key itself is the problem: keying on the full base64 string is memory-heavy (~5.8MB per 100s clip), and keying on a cheap hash of the base64 costs about the same as hashing the raw bytes (base64 is 1.33x larger). So unless profiling shows key-compute as a hotspot, which is unlikely given the codes cache eats the real cost, I'd leave the keyspace as is. Happy to add the memo as a follow-up if a workload surfaces where it matters.

File: sglang_omni/models/moss_tts_local/stages.py (hunk 5 above)
Comment (JiaxinD) [reply sglang_omni/models/moss_tts_local/stages.py:252]: Good catch — fixed in `2855ac1`. `CachedReferenceEncoder.__init__` now raises `ValueError` for non-positive `max_items`/`max_bytes` (fail-fast at construction, before any request can hit the empty-`OrderedDict` eviction path you described), with a regression test `test_cached_reference_encoder_rejects_nonpositive_capacity`.

File: sglang_omni/models/moss_tts_local/stages.py (hunk 2 above)
Comment (JiaxinD) [reply sglang_omni/models/moss_tts_local/stages.py:379]: Done in `2855ac1`. Extracted `_cached_encode(key, encode_fn, *, desc, revalidate)` holding the whole single-flight skeleton (lock -> hit / inflight-follower / leader, follower wait, leader encode + put). `encode()` and `encode_data_uri()` now only compute the key and pass an `encode_fn` callable; `encode()` additionally passes a `revalidate()` for its TOCTOU re-stat (data-URI needs none, so it omits it). All concurrency / failure-isolation / clone tests stay green.

File: sglang_omni/models/moss_tts_local/stages.py (hunk 3 above)
Comment (JiaxinD) [reply sglang_omni/models/moss_tts_local/stages.py:355]: Done in `2855ac1` -- pulled the cadence out as a class attribute `CachedReferenceEncoder.LOG_INTERVAL_S` (= 60.0), so it is easy to tune or override.

File: sglang_omni/models/moss_tts_local/stages.py (hunk 1 above)
Comment (JiaxinD) [reply sglang_omni/models/moss_tts_local/stages.py:396]: The flow is already validate-then-decode: `_DATA_URI_RE.match(ref_audio)` runs first and the `if match is None: raise ValueError(...)` guard rejects non-data-URI refs, so `base64.b64decode(...)` only ever sees a matched payload and `raw` is the decoded bytes we hash. Were you thinking of a different invalid case (e.g. a matched header but corrupt base64)? That currently raises from `b64decode` in the leader and fails just that request.

Comment (JiaxinD) [reply sglang_omni/models/moss_tts_local/stages.py:396]: You're right, and the asymmetry is intentional -- thanks for catching the doc mismatch. File-path refs gate on duration first (`torchaudio.info` is a cheap metadata read; no point registering inflight for an oversized file). Data-URI refs run the duration check inside the leader (after inflight registration) for exactly the reason you gave: it needs `sf.read()` of the decoded buffer, so doing it once in the leader lets concurrent same-payload requests merge instead of each re-reading a large payload. Added a `Note` comment at the data-URI leader in `2855ac1`, and I'll fix the PR-description wording.

File: sglang_omni/models/moss_tts_local/stages.py (hunk 4 above)
Comment (JiaxinD) [reply sglang_omni/models/moss_tts_local/stages.py]: On the protection chain first: the memo key is `(size, mtime_ns, ctime_ns)`, and `ctime` updates on essentially any write -- so a normal overwrite of a scratch file like `/tmp/ref.wav` **does** invalidate (new ctime -> memo miss -> re-hash). The only residual stale window is a replacement that preserves size **and** mtime_ns **and** ctime_ns at once, reachable only by rolling the clock back during the write -- not a normal operation.

That said, you're both right to be cautious. The sentinel read this skips is microsecond-level, and the real cost (the ~0.25 s codec encode) is already absorbed by the codes cache -- so the optimization doesn't earn even a small stale-read doubt. I've defaulted the MOSS call sites back to `trust_stat=False` (`2855ac1`), keeping the sentinel-validated path; the `trust_stat` flag and fast path stay in `reference_path_cache_key` as an opt-in for deployments that can guarantee immutable references. Thanks both.

**Review by Ccyest — APPROVED** [review]
